orca-core/src/test/groovy/com/netflix/spinnaker/orca/pipeline/persistence/ExecutionRepositoryUtilSpec.groovy

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.orca.pipeline.persistence
+
+import spock.lang.Specification
+
+import static com.netflix.spinnaker.orca.pipeline.model.SyntheticStageOwner.STAGE_AFTER
+import static com.netflix.spinnaker.orca.pipeline.model.SyntheticStageOwner.STAGE_BEFORE
+import static com.netflix.spinnaker.orca.test.model.ExecutionBuilder.orchestration
+import static com.netflix.spinnaker.orca.test.model.ExecutionBuilder.stage
+
+class ExecutionRepositoryUtilSpec extends Specification {
+
+  def "should order stages by ref id"() {
+    given:
+    def one = stage {
+      type = "three"
+      refId = "3"
+    }
+    def two = stage {
+      type = "one"
+      refId = "1"
+    }
+    def three = stage {
+      type = "two"
+      refId = "2"
+    }
+    def execution = orchestration {}
+
+    when:
+    ExecutionRepositoryUtil.sortStagesByReference(execution, [three, one, two])
+
+    then:
+    execution.stages*.refId == ["1", "2", "3"]
+  }
+
+  def "should order stages by synthetic owner"() {
+    given:
+    def execution = orchestration {
+      stage {
+        type = "findImage"
+        refId = "1"
+      }
+      stage {
+        id = "2"
+        type = "deploy"
+        refId = "2"
+        requisiteStageRefIds = ["1"]
+      }
+      stage {
+        type = "createServerGroup"
+        refId = "2<1"
+        syntheticStageOwner = STAGE_BEFORE
+        parentStageId = "2"
+      }
+      stage {
+        type = "applySourceServerGroupCapacity"
+        refId = "2<1>2"
+        syntheticStageOwner = STAGE_AFTER
+        requisiteStageRefIds = ["2<1>1"]
+        parentStageId = "2"
+      }
+      stage {
+        type = "disableCluster"
+        refId = "2<1>1<1"
+        syntheticStageOwner = STAGE_BEFORE
+        parentStageId = "2"
+      }
+      stage {
+        type = "shrinkCluster"
+        refId = "2<1>1"
+        syntheticStageOwner = STAGE_AFTER
+        parentStageId = "2"
+      }
+    }
+
+    def stages = new ArrayList(execution.stages)
+    execution.stages.clear()
+
+    when:
+    ExecutionRepositoryUtil.sortStagesByReference(execution, stages)
+
+    then:
+    execution.stages*.refId == ["1", "2<1", "2<1>1<1", "2", "2<1>2", "2<1>1"]
```

Comment:
Why do you expect `2<1>2` and `2<1>1` to come _after_ `2`? That doesn't make sense to me.

Reply:
Well, I don't know. I'm just going off of what the current implementation does. Probably the worst kind of test one could have.

Reply:
Yeah I think this ordering reflects the (IMO incorrect) parent stage ids in the fixtures.

Reply:
In a real pipeline I'd expect the ordering to be `["1", "2<1", "2<1>1<1", "2<1>1", "2<1>2", "2"]`